### ISSUE TITLE
NoClassDefFoundError 

### ISSUE THREAD

**Issue by alexxsanchezm** [hivemq/hivemq-mqtt-client#332]
## Expected behavior
The background service must complete successfully a connection with the remote broker. Ad this solution should work withe all devices this app supports.

## Actual behavior
Background process crash after try to connect

## To Reproduce
1. Start the service by host activity
2. Bind to the service
3. Initialize the mqtt client.
4. Perform a connection and using a CompletableScenario
### Reproducer code
```
    public void connect(StartData startData) {
        client = Mqtt5Client.builder()
                .serverHost(host)
                .serverPort(port)
                .identifier(clientId)
                .addConnectedListener(new MqttClientConnectedListener() {
                    @Override
                    public void onConnected(MqttClientConnectedContext context) {
                        Log.d(TAG, "On connected... " + context.getClientConfig().getServerHost());
                        //_listener.onClientConnected();
                    }
                })
                .addDisconnectedListener(new MqttClientDisconnectedListener() {
                    @Override
                    public void onDisconnected(MqttClientDisconnectedContext context) {
                        Log.d(TAG, "On disconnected... " + context.getCause());
                        //_listener.onConnectionBroken();
                    }
                })
                .automaticReconnectWithDefaultConfig()
                .buildRx();

        Log.d(TAG, "Performing connect...");
        this.connectState = ConnectState.CONNECTING;
        Mqtt5Connect connect = Mqtt5Connect.builder()
                .willPublish()
                    .topic(willTopic)
                .applyWillPublish()
                .build();

        Completable connectScenario = client.connect(connect)
                .doOnSuccess(this::connectSuccess)
                .doOnError(this::connectionError)
                .ignoreElement();

        Single<Mqtt5PublishResult> publishConnect
                = client.publish(Flowable.just(
                        Mqtt5Publish.builder()
                                .topic("d/" + this.clientId + START)
                                .payload(startData.toByteArray())
                                .build())).singleOrError();

        connectScenario
                .andThen(publishConnect)
                .subscribe(this::onConnectSuccess, this::connectionError);
    }
```
```
## Error log
 E  java.lang.NoClassDefFoundError: com.hivemq.client.internal.mqtt.message.connect.-$$Lambda$2ccuY8cM0LuLnZuGJfwaG9Phxz at com.hivemq.client.internal.mqtt.message.connect.MqttConnectBuilder.willPublish(MqttConnectBuilder.java:124)
at com.hivemq.client.internal.mqtt.message.connect.MqttConnectBuilder$Default.willPublish(MqttConnectBuilder.java:141)
```

## Details
- Affected HiveMQ MQTT Client version(s): 1.1.2
- Used JVM version: 
sourceCompatibility JavaVersion.VERSION_1_8
targetCompatibility JavaVersion.VERSION_1_8
- Android: 5.1

## Other details
In most of the devices the client connects, but not those running android 5.1. Is this version compatible with android 5.1?

**Comment by SgtSilvio** [comment]
Hi @alexxsanchezm 
You have to use the [Android RetroFix gradle plugin](https://github.com/SgtSilvio/android-retrofix) when using the HiveMQ MQTT Client on Android below API level 24/Android version 7.0:
https://hivemq.github.io/hivemq-mqtt-client/docs/installation/android.html#android-api-levels-below-24

**Comment by alexxsanchezm** [comment]
Hi @SgtSilvio,
Thanks to point me to Retrofix, but I wonder if there is a known conflict by using Protobuf generated java classes and Retrofix all together. In my requirements the data transported is serialized and unserialized by protobuf.
I've been doing several experiments before submit this message, but I'm getting this Class no found error (**java.lang.NoClassDefFoundError:**) every time I try to build a message to send out. Things I've tested so far: Changed the protobuf runtime version, regenerate protobuf classes, switched from protobuf lite back and forth.
But the only thing I've tested successfully is to disable the retrofix pluging and remove its dependencies.
I've really considering to not support andorid below 24-.  Thanks.

**Comment by SgtSilvio** [comment]
I do not know of a conflict of Protobuf and RetroFix.
I tried to use both in an Android project and it seems to work for me on Android 5.0.
Here is how I added the protobuf gradle plugin to `app/build.gradle`:
```groovy
buildscript {
    repositories {
        google()
        gradlePluginPortal()
    }
    dependencies {
        classpath 'gradle.plugin.com.github.sgtsilvio.gradle:android-retrofix:0.2.1'
    }
}

buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'com.google.protobuf:protobuf-gradle-plugin:0.8.10'
    }
}

apply plugin: 'com.android.application'
apply plugin: 'com.github.sgtsilvio.gradle.android-retrofix'
apply plugin: 'com.google.protobuf'

...

dependencies {
    implementation 'com.hivemq:hivemq-mqtt-client:1.1.2'

    implementation 'net.sourceforge.streamsupport:android-retrostreams:1.7.1'
    implementation 'net.sourceforge.streamsupport:android-retrofuture:1.7.1'

    ...
}

// the following is copied from the protobuf-gradle-plugin readme
dependencies {
    // You need to depend on the lite runtime library, not protobuf-java
    compile 'com.google.protobuf:protobuf-javalite:3.8.0'
}

protobuf {
    protoc {
        artifact = 'com.google.protobuf:protoc:3.8.0'
    }
    generateProtoTasks {
        all().each { task ->
            task.builtins {
                java {
                    option "lite"
                }
            }
        }
    }
}
```

**Comment by alexxsanchezm** [comment]
Hi @SgtSilvio ,
Yes, I can confirm, or that's what I think. There is an error when protobuf deserialize a message to parse it into a byte array. The only case I can be sure so far is when using float data types within message structures, here is the error log.

`java.lang.NoSuchMethodError: No virtual method getFloat(Ljava/lang/Object;J)F in class Lsun/misc/Unsafe; or its super classes (declaration of 'sun.misc.Unsafe' appears in /system/framework/core-libart.jar)
        at com.google.protobuf.UnsafeUtil$JvmMemoryAccessor.getFloat(UnsafeUtil.java:625)
        at com.google.protobuf.UnsafeUtil.getFloat(UnsafeUtil.java:147)
        at com.google.protobuf.MessageSchema.isFieldPresent(MessageSchema.java:5765)
        at com.google.protobuf.MessageSchema.getSerializedSizeProto3(MessageSchema.java:2005)
        at com.google.protobuf.MessageSchema.getSerializedSize(MessageSchema.java:1411)
        at com.google.protobuf.AbstractMessageLite.getSerializedSizeInternal(AbstractMessageLite.java:116)`

**Comment by SgtSilvio** [comment]
I tried it with a float field and I get the same error, but I also get the error when removing the RetroFix plugin and its dependencies.
This is an issue of protobuf version 3.8.0: https://github.com/protocolbuffers/protobuf/issues/6305
It was resolved in protobuf version 3.10.0.
I updated the build.gradle above. Then it works for me.

**Comment by alexxsanchezm** [comment]
Yeah, definitely that was the problem. I just wasn't sure where this error came from.
I think this can be closed. Thank you.
